### PR TITLE
refactor(ivy): separate container into 2 instructions

### PR DESCRIPTION
**Pull request by kara**
This PR separates the `container` instruction into 2 instructions: `container` and `template`. 

Currently, we use `container()` to encompass two opposite use cases: 

1) containers for inline views (i.e. insertion points)

2) `<ng-template>` declarations (i.e. definition points)

This can get confusing because sometimes we're defining an `ng-template` in one view but inserting it somewhere else, and yet we're using containers for both the definition point *and* the insertion point. Two instructions makes the mental model a bit simpler.

Before:
```html
<ng-template #foo>
  <div></div>
</ng-template>
```

```ts
container(0, MyComponent_Template_0, null, [], ['foo', '']);
```

After:
```html
<ng-template #foo>
  <div></div>
</ng-template>
```

```ts
template(0, MyComponent_Template_0, null, [], ['foo', '']);
```

### CODE REVIEW

**Review by mhevery — APPROVED** [review]

**Comment by kara** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:208928506:BASE:208928556:1534390374135:391738cf)

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -1827,19 +1827,52 @@ export function createLContainer(
...
+ * % if (showing) {
+ *   <div></div>
+ * % }
+ *
+ * @param index The index of the container in the data array
+ */
 export function container(
-    index: number, template?: ComponentTemplate<any>| null, tagName?: string | null,
-    attrs?: TAttributes | null, localRefs?: string[] | null): void {
+    index: number, tagName?: string | null, attrs?: TAttributes | null): void {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: We should probably get rid of `tagName` and  `attrs` as you can't have them on a JS block (`% if (showing) {`). We do have `tagName` and  `attrs` on containers for content projection purposes but IMO content projection of JS blocks is another topic that needs design.

tl;dr; I would vote for removing `tagName` and  `attrs`  from the `container` instruction corresponding to JS blocks.

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1827,19 +1827,52 @@ export function createLContainer(
...
- * @param template Optional inline template
+ * @param templateFn Inline template
  * @param tagName The name of the container element, if applicable
  * @param attrs The attrs attached to the container, if applicable
  * @param localRefs A set of local reference bindings on the element.
  */
+export function template(
+    index: number, templateFn: ComponentTemplate<any>| null, tagName?: string | null,
+    attrs?: TAttributes | null, localRefs?: string[] | null) {
+  const node = containerInternal(index, tagName, attrs, localRefs);
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: To ponder: when I was considering this split I was asking myself a question: should we use `LContainerNode` for `<ng-template>`? In reality it is more like an pseudo-element node (similar to `<ng-container>` but without children).

I'm not saying that we should change this in this PR, just leaving a note so we can consider using another `LNode` type (probably would require a new node type, sth like `LTemplateNode`).

File: packages/core/src/render3/instructions.ts [hunk 3]
```diff
@@ -1827,19 +1827,52 @@ export function createLContainer(
...
  *
  * @param index The index of the container in the data array
- * @param template Optional inline template
+ * @param templateFn Inline template
  * @param tagName The name of the container element, if applicable
  * @param attrs The attrs attached to the container, if applicable
  * @param localRefs A set of local reference bindings on the element.
  */
+export function template(
+    index: number, templateFn: ComponentTemplate<any>| null, tagName?: string | null,
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions.ts]: To consider: should we generate an empty function for cases where `<ng-template>` has no children? It often happens where `<ng-template>` is used as an insertion point (ex.:  `<ng-template [ngTemplateOutlet]="tplRef"></ng-template>`)

File: packages/core/src/render3/instructions.ts (hunk 1 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: Yeah, that was my understanding too, so that's what I originally had. But it turns out we have a test for projection in inline views here: https://github.com/angular/angular/blob/5f69fe67d016552e88ccea6bd25dd71537895132/packages/core/test/render3/content_spec.ts#L1792.

So I tried to keep the behavior as is pending discussion. Given your response though, sounds like maybe the test should have been `ngIf`?

File: packages/core/src/render3/instructions.ts (hunk 2 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: Agreed. I can leave a TODO for further discussion.

File: packages/core/src/render3/instructions.ts (hunk 1 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: Looking at this more, I think the test is just wrong and should be `ngIf`. Updated it.

Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions.ts]: > Looking at this more, I think the test is just wrong and should be ngIf. Updated it.

Yep, totally. The test was written before we had `ViewContainerRef` / `NgIf`.
